doc/rozofsmount.8
.\" Process this file with
.\" groff -man -Tascii rozofsmount.8
.\"
.TH ROZOFSMOUNT 8 "APRIL 2014" RozoFS "User Manuals"
.SH NAME
rozofsmount \- RozoFS fuse mount
.SH SYNOPSIS
.B rozofsmount
[\fB\-f\fP]
[\fB\-H\fP \fIEXPORT_HOST\fP]
[\fB\-E\fP \fIEXPORT_PATH\fP]
[\fB\-P\fP \fIEXPORT_PASSWD\fP]
[\fB\-o\fP \fIopt\fP[,\fIopt\fP]...]
\fImountpoint\fP
.PP
.B rozofsmount
\fB\-h\fP|\fB\-\-help\fP
.PP
.B rozofsmount
\fB\-V\fP|\fB\-\-version\fP
.PP
.SH DESCRIPTION
.B rozofsmount
mount a 
.BR rozofs (7) 
filesystem exported by
.BR exportd (8)
.SH OPTIONS

.SS "General options:"
.PP
.TP
\fB\-h\fP, \fB\-\-help
Display help and exit.
.TP
\fB\-V\fP, \fB\-\-version
Display version information and exit.
.PP
.SS "FUSE options:"
.TP
\fB\-f\fP
Foreground operation.
.PP
.SS "RozoFS options:"
.PP
.TP
\fB\-H\fP \fIEXPORT_HOST\fP, \fB\-o exporthost=\fP\fIEXPORT_HOST\fP
List of '/' separated addresses (or dns names) where exportd daemon is running (default is rozofsexport). When securing the exportd thanks to a virtual IP address only the virtual IP address is declared (i.e "192.168.10.1"). When no virtual IP address is used the list of "physical" IP addresses of each host securing the exportd must be given (i.e "192.168.10.1/192.168.20.1"). 
.TP
\fB\-E\fP \fIEXPORT_PATH\fP, \fB-o exportpath=\fP\fIEXPORT_PATH\fP
Root path of an export see
.BR exportd (8)
for further details. (default is /srv/rozofs/exports/export).
.TP
\fB\-E\fP \fIEXPORT_PASSWD\fP, \fB\-o exportpasswd=\fP\fIEXPORT_PASSWD\fP
Address (or dns name) where exportd daemon is running (default is empty).
.TP
\fB\-o rozofsbufsize=\fP\fIN\fP
Specify size of I/O buffer in KiB (in range: 128..8192 - default: 256).
.TP
\fB\-o rozofsminreadsize=\fP\fIN\fP
Specify minimum read size on disk in KiB (default value is same as RozoFS block size).
.TP
\fB\-o rozofsmaxwritepending=\fP\fIN\fP
Specify the number of write request(s) that can be sent for an open file from the rozofsmount toward the storcli asynchronously (default: 4).
.TP
\fB\-o rozofsmaxretry=\fP\fIN\fP
Specify number of retries before I/O error is returned (default: 50).
.TP
\fB\-o rozofsexporttimeout=\fP\fIN\fP
Specify timeout in seconds for exportd requests (default: 25).
.TP
\fB\-o rozofsstoragetimeout=\fP\fIN\fP
Specify timeout in seconds for IO storaged requests (default: 4).
.TP
\fB\-o rozofsstorclitimeout=\fP\fIN\fP
Specify timeout in seconds for storcli requests (default: 15).
.TP
\fB\-o rozofsattrtimeout=\fP\fIN\fP
Specify timeout in seconds for which file/directory attributes are cached. Should not be set along with -o rozofsattrtimeoutms.
.TP
\fB\-o rozofsattrtimeoutms=\fP\fIN\fP
Specify timeout in milliseconds for which file/directory attributes are cached. This parameter is ignored if rozofsattrtimeout is also set.
.TP
\fB\-o rozofsentrytimeout=\fP\fIN\fP
Specify timeout in seconds for which name lookups will be cached. Should not be set along with -o rozofsentrytimeoutms.
.TP
\fB\-o rozofsentrytimeoutms=\fP\fIN\fP
Specify timeout in milliseconds for which name lookups will be cached. This parameter is ignored if rozofsentrytimeout is also set.
.TP
\fB\-o rozofssymlinktimeout=\fP\fIN\fP
Specify timeout in milliseconds for which symbolic targets will be cached (default: 1000ms).
.TP
\fB\-o debug_port=\fP\fIN\fP
Specify the base debug port for rozofsmount (default: none).
.TP
\fB\-o instance=\fP\fIN\fP
Specify rozofsmount instance number (default: 0).
.TP
\fB\-o rozofscachemode=\fP\fIN\fP
Specify the cache mode: 0: no cache, 1: direct_io, 2: keep_cache (default: 0).
.TP
\fB\-o rozofsmode=\fP\fIN\fP
Specify the operating mode of rozofsmount: 0: filesystem, 1: block mode (default: 0).
.TP
\fB\-o rozofsnbstorcli=\fP\fIN\fP
Specify the number of storcli process(es) to use (in range: 1..2 - default: 1).
.TP
\fB\-o rozofsshaper : Deprecated !
.TP
\fB\-o rozofsrotate=\fP\fIN\fP
Specify the modulo on read distribution rotation (default: 0).
.TP
\fB\-o posixlock : Deprecated !
.TP
\fB\-o bsdlock : Deprecated !
.TP
\fB\-o nolock
To disable BSD as well as POSIX file lock.
.TP
\fB\-o noXattr
Disable support of extended attributes.
.TP
\fB\-o mojThreadWrite=\fP\fI1|0\fP
Enable|disable the use of the Mojette threads in the STORCLI for writes.
.TP
\fB\-o mojThreadRead=\fP\fI1|0\fP
Enable|disable the use of the Mojette threads in the STORCLI for reads.
.TP
\fB\-o mojThreadThreshold=\fP\fIbytes\fP
Modify the threshold in number of bytes above which the Mojette threads in the STORCLI are used.
.TP
\fB\-o writerOnly
Tells the client that it is the only writter on the files it writes, and so needs not to re-read the last block when extending a file.
.TP
\fB\-o localPreference
Favor local storage on read to save network bandwidth in case of poor network connection.
.PP
.SH "REPORTING BUGS"
Report bugs to <bugs@fizians.org>.
.SH COPYRIGHT
Copyright (c) 2013 Fizians SAS. <http://www.fizians.com>

Rozofs is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published
by the Free Software Foundation, version 2.

Rozofs is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
.SH AUTHOR
Fizians <http://www.fizians.org>
.SH "SEE ALSO"
.BR rozofs (7),
.BR exportd (8),
.BR storaged (8),
.BR geomgr (8)


debian/changelog
rozofs (1.6.alpha4) stable; urgency=low

  * (rozofs) See git log for changes

 -- Fizians S.A.S. <devel@rozofs.com>  Mon, 01 Jan 2014 00:00:00 +0100


manager/rozofs/core/rozofsmount.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2013 Fizians SAS. <http://www.fizians.com>
# This file is part of Rozofs.
#
# Rozofs is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation, version 2.
#
# Rozofs is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

from rozofs.core.agent import Agent, ServiceStatus
from rozofs.core.constants import ROZOFSMOUNT_MANAGER
import os
import subprocess
from rozofs.ext.fstab import Fstab, Line

class RozofsMountConfig(object):
    def __init__(self, export_host, export_path, instance, mountpoint=None, options=None):

        self.export_host = export_host
        self.export_path = export_path
        self.instance = instance

        if mountpoint is None:
            self.mountpoint = os.path.join('/mnt/rozofs@%s' % export_host.replace('/','-'), export_path.split('/')[-1])
        else:
            self.mountpoint = mountpoint
        if options is None:
            self.options = []
        else:
            self.options = options

    def __eq__(self, other):
        return self.export_host == other.export_host and self.export_path == other.export_path and self.mountpoint == other.mountpoint

class RozofsMountAgent(Agent):
    '''
    Managed rozofsmount via fstab
    '''

    __FSTAB = '/etc/fstab'
    __MTAB = '/etc/mtab'
    __FSTAB_LINE = "rozofsmount\t%s\trozofs\texporthost=%s,exportpath=%s,instance=%d%s,_netdev\t0\t0\n"

    def __init__(self, mountdir='/mnt'):
        """
        ctor
        @param mountdir : the parent directory name for rozofs mount points
        """
        Agent.__init__(self, ROZOFSMOUNT_MANAGER)
        self._mountdir = mountdir

    def _mount_path(self, export_host, export_path):
        return os.path.join(self._mountdir, "rozofs@%s" % export_host.replace('/','-'),
                            export_path.split('/')[-1])

    #
    # mount points management
    #
    def _mount(self, path):
        cmds = ['mount', path]
        with open('/dev/null', 'w') as devnull:
            p = subprocess.Popen(cmds, stdout=devnull, stderr=subprocess.PIPE)
            if p.wait() is not 0 :
                raise Exception(p.communicate()[1])
        return True

    def _umount(self, path):
        cmds = ['umount', path]
        with open('/dev/null', 'w') as devnull:
            p = subprocess.Popen(cmds, stdout=devnull, stderr=subprocess.PIPE)
            if p.wait() is not 0 :
                raise Exception(p.communicate()[1])
        return True

    def _list_mount(self):
        with open(self.__MTAB, 'r') as mtab:
            mount = [l.split()[1] for l in mtab if l.split()[0] == "rozofs"]
        return mount

#    def _is_mount(self, share):
#        return self._mount_path(share) in self._list_mount()

    def _add_mountpoint(self, config, instance):
        fstab = Fstab()
        fstab.read(self.__FSTAB)

        mount_path = self._mount_path(config.export_host, config.export_path)

        # Create mountpoint
        if not os.path.exists(config.mountpoint):
            os.makedirs(config.mountpoint)
        else:
            if not os.path.isdir(config.mountpoint):
                raise Exception('mountpoint: %s is not a directory.'
                                 % config.mountpoint)
                # Check if empty
        # add a line to fstab
        if not config.options:
            stroptions=""
        else:
            stroptions = ',' + ','.join(config.options)
        
        fstab.lines.append(Line(self.__FSTAB_LINE % (config.mountpoint,
                                                     config.export_host,
                                                     config.export_path,
                                                     instance,
                                                     stroptions)))
        fstab.write(self.__FSTAB)

    def _remove_mountpoint(self, config):
        fstab = Fstab()
        fstab.read(self.__FSTAB)
        mount_path = self._mount_path(config.export_host, config.export_path)
        if os.path.exists(config.mountpoint):
            os.rmdir(config.mountpoint)
        # remove the line from fstab
        newlines = [l for l in fstab.lines if l.directory != config.mountpoint]
        fstab.lines = newlines
        fstab.write(self.__FSTAB)

    def _list_mountpoint(self):
        fstab = Fstab()
        fstab.read(self.__FSTAB)
        return [l.directory for l in fstab.get_rozofs_lines()]

    def get_service_config(self):
        fstab = Fstab()
        fstab.read(self.__FSTAB)
        configurations = []
        for l in fstab.get_rozofs_lines():
            o = l.get_rozofs_options()
            configurations.append(RozofsMountConfig(o["host"], o["path"],
                                                    o["instance"],
                                                    o["mountpoint"]))
        return configurations

    def set_service_config(self, configurations):
        responses = {}
        instance = 0
        currents = self.get_service_config()

        # find an instance
        if currents:
            instance = max([int(c.instance) for c in currents]) + 1

        # For added mountpoint(s)
        for config in [c for c in configurations if c not in currents]:
            mnt_statuses = {}
            mnt_statuses['config'] = True
            mnt_statuses['service'] = True
            # Add the mountpoint
            try:
                self._add_mountpoint(config, instance)
            except Exception as e:
                mnt_statuses['config'] = type(e)(str(e))
            # Mount it
            if mnt_statuses['config'] is True:
                try:
                    self._mount(config.mountpoint)
                except Exception as e:
                    mnt_statuses['service'] = type(e)(str(e))
            else:
                mnt_statuses['service'] = False

            # Add mnt statuses to response
            responses[config.mountpoint] = mnt_statuses
            instance = instance + 1

        # For removed mountpoint(s)
        for config in [c for c in currents if c not in configurations]:
            mnt_statuses = {}
            mnt_statuses['config'] = True
            mnt_statuses['service'] = True
            # Umount this mountpoint
            try:
                self._umount(config.mountpoint)
            except Exception as e:
                mnt_statuses['service'] = type(e)(str(e))
            # Remove it from fstab
            if mnt_statuses['service'] is True:
                try:
                    self._remove_mountpoint(config)
                except Exception as e:
                    mnt_statuses['config'] = type(e)(str(e))
            else:
                mnt_statuses['config'] = False
            # Add mnt statuses to response
            responses[config.mountpoint] = mnt_statuses

        return responses

    def get_service_status(self):
        statuses = {}
        for m in self._list_mountpoint():
            if m in self._list_mount():
                statuses[m] = ServiceStatus.STARTED
            else:
                statuses[m] = ServiceStatus.STOPPED
        return statuses

    def set_service_status(self, status):
        changes = {}

        if status == ServiceStatus.STARTED:
            # For each mountpoint defined in fstab
            for mp in self._list_mountpoint():
                # Check it's already mounted
                if mp in self._list_mount():
                    changes[mp] = False
                else:
                    changes[mp]= self._mount(mp)

        if status == ServiceStatus.STOPPED:
            # For each mountpoint defined in fstab
            for mp in self._list_mountpoint():
                # Check it's not mounted
                if mp not in self._list_mount():
                    changes[mp] = False
                else:
                    changes[mp]= self._umount(mp)

        return changes